webapp/src/ts/services/contact-save.service.ts

```diff
@@ -0,0 +1,191 @@
+})
+export class ContactSaveService {
+  private servicesActions;
+  private readonly CONTACT_FIELD_NAMES = [ 'parent', 'contact' ];
+
+  constructor(
+    private store:Store,
+    private contactTypesService:ContactTypesService,
+    private dbService:DbService,
+    private enketoTranslationService:EnketoTranslationService,
+    private extractLineageService:ExtractLineageService,
+  ) {
+    this.servicesActions = new ServicesActions(store);
+  }
+
+  private generateFailureMessage(bulkDocsResult) {
+    return _reduce(bulkDocsResult, (msg, result) => {
+      let newMsg = msg;
+      if (!result.ok) {
+        if (!newMsg) {
+          newMsg = 'Some documents did not save correctly: ';
+        }
+        newMsg += result.id + ' with ' + result.message + '; ';
+      }
+      return newMsg;
+    }, null);
+  }
+
+  private prepareSubmittedDocsForSave(original, submitted) {
+    const doc = this.prepare(submitted.doc);
+
+    return this
+      .prepareAndAttachSiblingDocs(submitted.doc, original, submitted.siblings)
+      .then((siblings) => {
+        const extract = item => {
+          item.parent = item.parent && this.extractLineageService.extract(item.parent);
+          item.contact = item.contact && this.extractLineageService.extract(item.contact);
+        };
+
+        siblings.forEach(extract);
+        extract(doc);
+
+        // This must be done after prepareAndAttachSiblingDocs, as it relies
+        // on the doc's parents being attached.
+        const repeated = this.prepareRepeatedDocs(submitted.doc, submitted.repeats);
+
+        return {
+          docId: doc._id,
+          preparedDocs: [ doc ].concat(repeated, siblings) // NB: order matters: #4200
+        };
+      });
+  }
+
+  // Prepares document to be bulk-saved at a later time, and for it to be
+  // referenced by _id by other docs if required.
+  private prepare(doc) {
+    if (!doc._id) {
+      doc._id = uuidV4();
+    }
+
+    if (!doc._rev) {
+      doc.reported_date = Date.now();
+    }
+
+    return doc;
+  }
+
+  private prepareRepeatedDocs(doc, repeated) {
+    const childData = repeated?.child_data || [];
+    return childData.map(child => {
+      child.parent = this.extractLineageService.extract(doc);
+      return this.prepare(child);
+    });
+  }
+
+  private extractIfRequired(name, value) {
+    return this.CONTACT_FIELD_NAMES.includes(name) ? this.extractLineageService.extract(value) : value;
+  }
+
+  // Mutates the passed doc to attach prepared siblings, and returns all
+  // prepared siblings to be persisted.
+  // This will (on a correctly configured form) attach the full parent to
+  // doc, and in turn siblings. See internal comments.
+  private prepareAndAttachSiblingDocs(doc, original, siblings) {
+    if (!doc._id) {
+      throw new Error('doc passed must already be prepared with an _id');
+    }
+
+    const preparedSiblings = [];
+    let promiseChain = Promise.resolve();
+
+    this.CONTACT_FIELD_NAMES.forEach(fieldName => {
```

Comment:
This is a very long forEach callback, can we extract it in functions by concerns (if that makes sense here) ? :)